### PR TITLE
tests: add rosdep_repo_check alpine support

### PR DESCRIPTION
**Pull request by russkel**
As mentioned in #30037, I said I would add Alpine Linux test support.

This adds support for Alpine Linux style `apk` packages.

### CODE REVIEW

**Comment by ivanpauno** [comment]
> I think this is 90% of the work yet I haven't found a good way to test it, so it hasn't been run yet.

I think you can try adding a new alpine key, either in a commit in this PR or in a new PR that also includes this changes.
We should test both a correct rosdep key and one that has an issue.

File: test/rosdep_repo_check/alpine.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    # o:rtpproxy
+    # m:Natanael Copa <ncopa@alpinelinux.org>
+    # t:1602354892
+    # c:183e99f73bb1223768aa7231a836a3e98e94c03e
+    # i:docs rtpproxy=2.1.1-r0
+
+    fio = TextIOWrapper(f)
+    while True:
+        entry = {}
+        while (l := fio.readline()) != '\n':
```
Comment (cottsay) [inline test/rosdep_repo_check/alpine.py:66]: The `TextIOWrapper` doesn't support un-seekable streams, it seems. Here's a suggestion that worked for me:
```suggestion
    while True:
        entry = {}
        while (l := f.readline().decode('utf-8')) not in ['', '\n']:
```

The corresponding import should also be dropped.

File: test/rosdep_repo_check/alpine.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            k, v = l.strip().split(':', 1)
+            entry[k] = v
+
+        yield entry
```
Comment (cottsay) [inline test/rosdep_repo_check/alpine.py:70]: The `while True` needs to be terminated:
```suggestion
        if entry:
            yield entry
        else:
            break
```

File: test/rosdep_repo_check/alpine.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    base_url = replace_tokens(base_url, os_code_name, os_arch)
+    # get main or community or testing from the second last component of the url
+    pkg_source = os.path.split(os.path.split(base_url)[0])[1]
+    apkindex_url = os.path.join(base_url, 'APKINDEX.tar.gz')
+    print('Reading Alpine package metadata from ' + apkindex_url)
+
+    with open_gz_url(apkindex_url) as f:
+        with tarfile.open(fileobj=f) as tf:
+            index = tf.extractfile("APKINDEX")
```
Comment (cottsay) [inline test/rosdep_repo_check/alpine.py:93]: This doesn't seem to work correctly to parse the stream of data coming down and tries to seek around like a file on disk. After some manipulation, I was able to get it working with this change:
```suggestion
        with tarfile.open(mode='r|', fileobj=f) as tf:
            index = None
            for ti in tf:
                if ti.name == 'APKINDEX':
                    index = tf.extractfile(ti)
                    break
            if index is None:
                raise RuntimeError('APKINDEX url did not contain an APKINDEX file')
```

File: test/rosdep_repo_check/alpine.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+            for index_entry in parse_apkindex(index):
+                pkg_name, pkg_version = index_entry['o'], index_entry['V']
+                pkg_filename = "%s-%s.apk" % (pkg_name, pkg_version)
```
Comment (cottsay) [inline test/rosdep_repo_check/alpine.py:97]: nit: consistent styling:
```suggestion
                pkg_filename = '%s-%s.apk' % (pkg_name, pkg_version)
```

**Review by cottsay — CHANGES_REQUESTED** [review]
Quickest way to test is to leave some uncommitted changes in your stage and run the nose test manually:
```
$ git diff
diff --git a/rosdep/python.yaml b/rosdep/python.yaml
index ec7ab7082..c906d16c2 100644
--- a/rosdep/python.yaml
+++ b/rosdep/python.yaml
@@ -5759,6 +5759,7 @@ python3-aio-pika-pip:
     pip:
       packages: [aio-pika]
 python3-aiohttp:
+  alpine: [py3-aiohttp]
   arch: [python-aiohttp]
   debian: [python3-aiohttp]
   fedora: [python-aiohttp]
$ python3 -m nose test/rosdep_repo_check/test_rosdep_repo_check.py -s
...
```

For confirmation that things worked as expected, you're looking for verbose output like this: `Package 'py3-aiohttp' for alpine edge on x86_64 was found: https://pkgs.alpinelinux.org/package/edge/community/x86_64/py3-aiohttp`

Two general comments on this PR:
1. Thanks for opening it! I'm happy to see support for additional platforms added. Please feel free to suggest updates to the documentation that might help make future contributions easier.
2. We need to stick to the intended use of the `PackageEntry` field `source_name`, which should represent the name of the source package that provides the given binary package. This may very well not be a relevant concept on Alpine, in which case it should be omitted so that it defaults to the value of `name`. This leaves an issue with extracting the repository name from the package URL to construct the dashboard link, but I think this change should make that work:
   ```diff
   diff --git a/test/rosdep_repo_check/__init__.py b/test/rosdep_repo_check/__init__.py
   --- a/test/rosdep_repo_check/__init__.py
   +++ b/test/rosdep_repo_check/__init__.py
   @@ -252,8 +252,9 @@ def get_package_link(config, pkg, os_name, os_code_name, os_arch):
        :returns: a URL to a dashboard or package file.
        """
        for dashboard in config.get('package_dashboards', ()):
   -        if dashboard['pattern'].match(pkg.url):
   -            return dashboard['url'].format_map({
   +        match = dashboard['pattern'].match(pkg.url)
   +        if match:
   +            return match.expand(dashboard['url']).format_map({
                    'binary_name': pkg.binary_name,
                    'name': pkg.name,
                    'os_arch': os_arch,
   diff --git a/test/rosdep_repo_check/alpine.py b/test/rosdep_repo_check/alpine.py
   --- a/test/rosdep_repo_check/alpine.py
   +++ b/test/rosdep_repo_check/alpine.py
   @@ -83,8 +83,6 @@ def enumerate_alpine_packages(base_url, os_name, os_code_name, os_arch):
        """
    
        base_url = replace_tokens(base_url, os_code_name, os_arch)
   -    # get main or community or testing from the second last component of the url
   -    pkg_source = os.path.split(os.path.split(base_url)[0])[1]
        apkindex_url = os.path.join(base_url, 'APKINDEX.tar.gz')
        print('Reading Alpine package metadata from ' + apkindex_url)
    
   @@ -96,7 +94,7 @@ def enumerate_alpine_packages(base_url, os_name, os_code_name, os_arch):
                    pkg_name, pkg_version = index_entry['o'], index_entry['V']
                    pkg_filename = "%s-%s.apk" % (pkg_name, pkg_version)
                    pkg_url = os.path.join(base_url, pkg_filename)
   -                yield PackageEntry(pkg_name, pkg_version, pkg_url, source_name=pkg_source)
   +                yield PackageEntry(pkg_name, pkg_version, pkg_url)
    
    
    def alpine_base_url(base_url):
   diff --git a/test/rosdep_repo_check/config.yaml b/test/rosdep_repo_check/config.yaml
   --- a/test/rosdep_repo_check/config.yaml
   +++ b/test/rosdep_repo_check/config.yaml
   @@ -47,8 +47,8 @@ package_dashboards:
      url: https://software.opensuse.org/package/{source_name}
    - pattern: !regular_expression .*//archive.ubuntu.com/ubuntu/.*
      url: https://packages.ubuntu.com/{os_code_name}/{binary_name}
   -- pattern: !regular_expression .*//dl-cdn.alpinelinux.org/alpine/.*
   -  url: https://pkgs.alpinelinux.org/package/{os_code_name}/{source_name}/x86_64/{binary_name}
   +- pattern: !regular_expression .*//dl-cdn.alpinelinux.org/alpine/[^/]+/([^/]+)/.*
   +  url: https://pkgs.alpinelinux.org/package/{os_code_name}/\1/x86_64/{binary_name}
    
    supported_versions:
      alpine:
   ```

I made a few other inline comments - please feel free to take suggestions verbatim or adapt them as you please.

Thanks!

File: test/rosdep_repo_check/alpine.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def replace_tokens(string, os_code_name, os_arch):
+    """Replace tokens in the repository base URL."""
+    for key, value in {
+        '$basearch': os_arch,
+        '$releasever': os_code_name,
+    }.items():
+        string = string.replace(key, value)
+    return string
```
[the remark this answers is not in the document]
Comment (russkel) [reply test/rosdep_repo_check/alpine.py:44]: Sorry, not sure what you're asking me to do here. Is this incorrect? I based it off the `rpm.py` script hence having the similar code. I imagined the style was more aligned with RPM than it was debian.

File: test/rosdep_repo_check/alpine.py (hunk 2 above)
Comment (russkel) [reply test/rosdep_repo_check/alpine.py]: Infinite loops are rarely good.

**Comment by russkel** [comment]
```
Run nosetests -s
Package 'automake' for alpine edge on x86_64 was found: https://pkgs.alpinelinux.org/package/edge/main/x86_64/automake

Error: Package 'autopoint' could not be found for alpine edge on x86_64
```
:woman_dancing: 

Thanks @cottsay for the review! I didn't mean to leave the first round debugging to you, but I forgot about this.
I committed all changes.

I misinterpreted the meaning of `source`. I thought it meant source repo rather than source package. Thanks for fixing that in the root script.

File: test/rosdep_repo_check/alpine.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cottsay) [reply test/rosdep_repo_check/alpine.py:44]: What I'm trying to say is that the goal isn't to align the style with RPM or deb, but rather alpine. The RPM base URL tokens are the same as you see in `/etc/yum.repos.d/fedora.repo`, and the base URLs for the debian repositories are similar to what you find in `/etc/apt/sources.list`.

Here's what `/etc/apk/repositories` looks like:
```
https://dl-cdn.alpinelinux.org/alpine/edge/main
https://dl-cdn.alpinelinux.org/alpine/edge/community
```

It looks like we can't avoid using a token for the release part (in this case, `edge`) but it appears that the rest of the URL (the arch-specific part) follows a known structure. I'm OK with borrowing `$releasever` from RPM to make this work, but I'd rather not deviate from the "native" representation of an alpine base URL any more than we need to.

tl;dr, I'd like to see this:
```
!alpine_base_url https://dl-cdn.alpinelinux.org/alpine/$releasever/main
```
...rather than this:
```
!alpine_base_url https://dl-cdn.alpinelinux.org/alpine/$releasever/main/$basearch
```

**Comment by cottsay** [comment]
So we have two existing package enumerators: `deb` and `rpm`. As I understand it, the name of this packaging format is actually `apk`. Would it be better to replace the use of `alpine` with `apk`?

Legend has it that the R in RPM originally stood for "Red Hat", but as RPM was adopted on non-Red-Hat platforms it came to be a recursive acronym instead. I'm not saying that this is going to happen to Alpine's package format but clearly there's precedent, so using `apk` over `alpine` might be more future-proof.

On the other hand Android's `.apk` format is pretty widely used, too, so it might get confusing.

I'd like to hear thoughts from @nuclearsandwich on the matter as well.

File: test/rosdep_repo_check/alpine.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                    index = tf.extractfile(ti)
+                    break
+            if index is None:
+                raise RuntimeError('APKINDEX url did not contain an APKINDEX file')
+
+            for index_entry in parse_apkindex(index):
+                pkg_name, pkg_version = index_entry['o'], index_entry['V']
+                pkg_filename = '%s-%s.apk' % (pkg_name, pkg_version)
+                pkg_url = os.path.join(base_url, pkg_filename)
+                yield PackageEntry(pkg_name, pkg_version, pkg_url)
```
Comment (cottsay) [inline test/rosdep_repo_check/alpine.py:105]: Earlier I said:
> ... the `PackageEntry` field `source_name` ... should represent the name of the source package that provides the given binary package. This may very well not be a relevant concept on Alpine...

I think that this concept actually does carry over to Alpine as the "origin" field. Actually, I think that the code presented here is only enumerating the "base" packages and is missing all of the "sub-packages".

It's worth noting that Alpine's packaging dashboard seems to work the same given the names of sub-packages, where Fedora's requires a source package name, so we don't actually need to populate the source package name here for things to work for Alpine. It should probably be present for completeness and future-proofing, though.

Even if we didn't populate the `source_name` field, we should definitely enumerate sub-packages since a rosdep rule could certainly refer to one.

```suggestion
                pkg_name, pkg_version, source_name = index_entry['P'], index_entry['V'], index_entry['o']
                pkg_filename = '%s-%s.apk' % (pkg_name, pkg_version)
                pkg_url = os.path.join(base_url, pkg_filename)
                yield PackageEntry(pkg_name, pkg_version, pkg_url, source_name=source_name)
```

**Comment by nuclearsandwich** [comment]
> I'd like to hear thoughts from @nuclearsandwich on the matter as well.

I'm in favor of apk, postmarketOS and Adelie Linux are both Alpine derivatives using apk. I'm not overly concerned about the collision with the Android format since that's a package file format rather than a repository format or utility.

File: test/rosdep_repo_check/alpine.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (russkel) [reply test/rosdep_repo_check/alpine.py]: Understood now. Fixed.

**Comment by russkel** [comment]
> So we have two existing package enumerators: `deb` and `rpm`. As I understand it, the name of this packaging format is actually `apk`. Would it be better to replace the use of `alpine` with `apk`?
> 
> Legend has it that the R in RPM originally stood for "Red Hat", but as RPM was adopted on non-Red-Hat platforms it came to be a recursive acronym instead. I'm not saying that this is going to happen to Alpine's package format but clearly there's precedent, so using `apk` over `alpine` might be more future-proof.
> 
> On the other hand Android's `.apk` format is pretty widely used, too, so it might get confusing.
> 
> I'd like to hear thoughts from @nuclearsandwich on the matter as well.

Agreed. Changes made.

**Comment by russkel** [comment]
@cottsay anything else that needs attention?

**Comment by ivanpauno** [comment]
@cottsay friendly ping

**Review by cottsay — APPROVED** [review]
Sorry, this dropped off my radar. Let's do it!

**Comment by nuclearsandwich** [comment]
Thanks so much for the contribution @russkel I hope this makes building out the alpine rosdep definitions more maintainable!

**Comment by russkel** [comment]
My pleasure. Thanks all.
